proposals/0034-tpu-vote-using-quic.md

```diff
@@ -0,0 +1,112 @@
+created: 2024-11-13
+development: 
+  - Anza - WIP
+  - Firedancer - Not started
+---
+
+## Summary
+
+Use QUIC for transporting TPU votes among Solana validators. This requires
+supporing receiving QUIC based vote TPU packets on the server side and sending
+QUIC-based TPU vote packets on the client side.
+
+
+## Motivation
+
+As timely vote credits are awarded to validators, they might be incentived to
+increase the TPU vote traffic to ensure their votes are received in a timely
+manner. This could cause congestions and impact overall TPU vote processing
+effectiveness. The concurrent UDP based TPU vote does not have any flow control
+mechanism.
+
+We propose to apply the pattern taken for TPU transaction processing to TPU vote
+processing -- by utlizing the flow control mechanism which were developed
+including built-in QUIC protocol level flow control, and application-level rate
+limiting on connections and packets.
+
+## Alternatives Considered
+
+There is no readily-available alternative to QUIC which addresses some of the
+requirements such as security (reliability when applying QOS), low latency and
+flow control. We could solve the security and flow control with TLS over TCP
+the concern is with the latency and head-of-line problems. We could also
+customize and build our own rate limiting mechanism based on the UDP directly,
+this is non-trivial and cannot solve the security problem without also rely on
+some sort of crypto handshaking. 
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+On the server side, the validator will bind to a new QUIC endpoint. Its
+corresponding port will be published to the network in the ContactInfo via
+Gossip. The client side will use the TPU vote QUIC port published by the server
+to connect to the server.
+
+The TPU vote will be using the same QUIC implementation used by regular
+transaction transportation. The client and server both uses their validator's
+identity key to sign the certificate which is used to validate the validator's
+identity especially on the server side for the purpose of provding QOS based on
+the client's stakes by checking the client's Pubkey -- stake weighted QOS.
+
+Once a QUIC connection is established, the client can send vote transaction
+using QUIC UNI streams. In this design, a stream is used to send one single Vote
+transaction. After that the stream is closed.
+
+The server only supports connections from the nodes which has stakes who can
+vote. Connections from unstaked nodes are rejected with `disallowed` code.
+
+The following QOS mechanisms are employed:
+
+* Connection Rate Limiting from all clients
+* Connection Rate Limiting from a particular IpAddress
+* Total concurrent connections from all clients -- this is set to 2500
+* Max concurrent connections from a client Pubkey -- this is set to 1 for votes.
+* Max concurrent streams per connection -- this is allocated based on the ratio
+of the validator's stake over the total stakes of the network.
```

Comment:
What is the reasoning here? Why does higher-staked validator need more concurrent streams per connection?

Reply:
Give them more bandwidth as reward and ensure they have better chance landing their votes?